### PR TITLE
fix(metric): fix hummock table stats

### PR DESCRIPTION
**Pull request by zwang28**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

- Fix bug when calculating table stats change in compaction, from https://github.com/risingwavelabs/risingwave/pull/6390
- Remove stale_key_count stats from per table stat, because it's a per SST stat, so the aggregated per table one is inaccurate.
- Add grafana metric for per table stats.

Screen Shot 2022-11-23 at 2 07 12 PM

## Checklist

~~- [ ] I have written necessary rustdoc comments~~
~~- [ ] I have added necessary unit tests and integration tests~~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

File: src/storage/src/hummock/compactor/mod.rs [hunk 1]
```diff
@@ -597,6 +591,16 @@ impl Compactor {
             } else {
                 local_stats.skip_multi_version_key_count += 1;
             }
+
+            if last_table_id.is_none()
```
Comment (soundOfDestiny) [inline src/storage/src/hummock/compactor/mod.rs:595]: We can use `map_or` here

**Review by soundOfDestiny — APPROVED** [review]

**Comment by hzxa21** [comment]
Just to confirm, can we check "Real Process Time" under "Hummock Manager" section to see whether there is a non-neligible increase in the real processing time of `commit_epoch` and `report_compact_task` ?

**Comment by zwang28** [comment]
> Just to confirm, can we check "Real Process Time" under "Hummock Manager" section to see whether there is a non-neligible increase in the real processing time of `commit_epoch` and `report_compact_task` ?

I continuously create new MV (nexmark_q7 with different MV name) in a 3 CN cluster. 

- The `real processing time` of commit_epoch and report_compact_task don't increase.
- At most 200 table id in a SST observed, which means sizeof the single row table stats KV written to meta store is about 6kB.
- Then a CN timeout when reading object store. Maybe because CN is overloaded or some object store quota exceeded?
`meta-node-0  | 2022-11-23T13:19:58.259449Z  WARN risingwave_meta::barrier: Failed to commit epoch 3406188174835712: Rpc error: gRPC error (Internal error): Actor 196 exit unexpectedly: Executor error: Storage error: Hummock error: ObjectStore failed with IO error Internal error: read "hummock_001/94/1095.data" in block Some(BlockLocation { offset: 6291618, size: 1048618 }) failed, error: timeout: error trying to connect: HTTP connect timeout occurred after 3.1s`
- So I haven't yet verified whether more table (200+) will affect `real processing time` significantly.
- I think the most likely reason related to table stats that may increase `real processing time` is the size of single row table stats KV. So I will do another test that manually enlarge the size, by adding dummy field.

**Comment by hzxa21** [comment]
> > Just to confirm, can we check "Real Process Time" under "Hummock Manager" section to see whether there is a non-neligible increase in the real processing time of `commit_epoch` and `report_compact_task` ?
> 
> I continuously create new nexmark_q7 with different MV names in a 3 CN cluster, to achieve large number of tables.
> 
> - The `real processing time` of commit_epoch and report_compact_task don't increase, until a CN actor fails.
> - At most 200 table id in a SST observed, which means sizeof the single row table stats KV written to meta store is about 6kB.
> - Then a CN actor fails due to object store read timeout. Maybe because CN is overloaded or some object store quota exceeded?
> `meta-node-0  | 2022-11-23T13:19:58.259449Z  WARN risingwave_meta::barrier: Failed to commit epoch 3406188174835712: Rpc error: gRPC error (Internal error): Actor 196 exit unexpectedly: Executor error: Storage error: Hummock error: ObjectStore failed with IO error Internal error: read "hummock_001/94/1095.data" in block Some(BlockLocation { offset: 6291618, size: 1048618 }) failed, error: timeout: error trying to connect: HTTP connect timeout occurred after 3.1s`
> - So I haven't yet verified whether more table (200+) will affect `real processing time` significantly.
> - I think the most likely reason related to table stats that may increase `real processing time` is the size of single row table stats KV. So I will do another test that manually enlarge the size, by adding dummy field.

Thanks for the effort. We haven't seen "HTTP connect timeout occurred after 3.1s" before. How many MVs did you create? Did the recovery process succeed or the EC2 just died?

**Comment by zwang28** [comment]
> How many MVs did you create? Did the recovery process succeed or the EC2 just died?

https://github.com/risingwavelabs/risingwave/issues/6551

**Comment by zwang28** [comment]
> another test that manually enlarge the size, by adding dummy field.

We can see the baseline is 2.5ms (no dummy filed added).
Large table number deteriorates meta store write latency:

- 1k table, ~32kB -> 2.5ms. I think it equals to baseline because the metric's minimal accuracy is 2.5ms.
- 10k table -> 7.5ms
- 20k table -> 17.5ms
- 100k table -> 75ms

Screen Shot 2022-11-24 at 2 15 30 PM

**Comment by hzxa21** [comment]
> > another test that manually enlarge the size, by adding dummy field.
> 
> We can see the baseline is 2.5ms (no dummy table stats entry added). Large table number deteriorates meta store write latency:
> 
> * 1k table, table stats size ~32kB -> 2.5ms. I think it equals to baseline because the metric's minimal accuracy is 2.5ms.
> * 10k table -> 7.5ms
> * 20k table -> 17.5ms
> * 100k table -> 75ms
> 
> Screen Shot 2022-11-24 at 2 15 30 PM
> 
> As long as the total table number is ~1k (about 100 nexmark_q7 MVs), the affection of table stats is negligible.

Clear enough. Thanks! 

I think it is fine as long as the latency is below 10ms. We can keep it as it is and optimize it when we have the need to support more than 10k tables in one cluster, which I think there can be more issues to solve in addition to optimizing the table stats.

**Pull request merged**